packages/portfolio-contract/src/durable-queue.exo.ts

```diff
@@ -0,0 +1,129 @@
+/**
+ * @file Durable queue helper backed by a durable map and index pointers.
+ */
+import { Fail } from '@endo/errors';
+import { M } from '@endo/patterns';
+import type { MapStore, StoreOptions } from '@agoric/store';
+import type { Zone } from '@agoric/zone';
+
+export type DurableQueueState<V> = {
+  readonly entries: MapStore<bigint, V>;
+  head: bigint;
+  tail: bigint;
+};
+
+export type DurableQueue<V> = ReturnType<
+  ReturnType<typeof prepareDurableQueue<V>>
+>;
+
+export const prepareDurableQueue = <V>(
+  zone: Zone,
+  queueName: string,
+  options: StoreOptions = {},
+) => {
+  const collections = zone.detached();
+
+  const indicesShape = M.splitRecord(
+    {
+      head: M.bigint(),
+      tail: M.bigint(),
+    },
+    {},
+  );
+
+  const entryShape = M.splitRecord(
+    {
+      index: M.bigint(),
+      value: options.valueShape || M.any(),
+    },
+    {},
+  );
+
+  const apiShape = options.valueShape || M.any();
+
+  return zone.exoClass(
+    queueName,
+    M.interface(queueName, {
+      enqueue: M.call(apiShape).returns(M.undefined()),
+      dequeue: M.call().returns(M.opt(apiShape)),
+      peek: M.call().returns(M.opt(apiShape)),
+      isEmpty: M.call().returns(M.boolean()),
+      size: M.call().returns(M.number()),
+      clear: M.call().returns(M.undefined()),
+      getIndices: M.call().returns(indicesShape),
+      entries: M.call().returns(M.arrayOf(entryShape)),
+    }),
+    (
+      storeName: string,
+      { head = 0n, tail = head }: { head?: bigint; tail?: bigint } = {},
+    ): DurableQueueState<V> => {
+      tail >= head || Fail`initial tail ${tail} must be >= head ${head}`;
+      const entries = collections.mapStore<bigint, V>(storeName, {
+        keyShape: M.bigint(),
+        ...options,
+      });
+      return {
+        entries,
+        head,
+        tail,
+      };
+    },
+    {
+      enqueue(value: V) {
+        const { state } = this;
+        state.entries.init(state.tail, value);
+        state.tail += 1n;
+      },
+      dequeue() {
+        const { state } = this;
```

Comment:
It's a little easier to tell reading from writing the `state` record if we do all the reads at once:

```suggestion
        const { head, tail, entries } = this.state;
```